models/event_tickets.py

```diff
@@ -0,0 +1,190 @@
+    state = db.Column(db.String, nullable=False)
+    proposal_id = db.Column(db.Integer, db.ForeignKey("proposal.id"), nullable=False)
+    user_id = db.Column(db.Integer, db.ForeignKey("user.id"), nullable=False)
+    rank = db.Column(db.Integer, nullable=True)
+    ticket_count = db.Column(db.Integer, nullable=False, default=1)
+    ticket_codes = db.Column(db.String, nullable=True)
+
+    def __init__(self, user_id, proposal_id, state, ticket_count=1, rank=None):
+        if state not in EVENT_TICKET_STATES:
+            raise EventTicketException(f"invalid ticket state {state}")
+
+        if state == "entered-lottery" and ((rank and rank < 0) or rank is None):
+            raise EventTicketException("rank must be greater than or equal to 0")
+
+        if ticket_count <= 0:
+            raise EventTicketException("ticket_count must be greater than 0")
+
+        self.user_id = user_id
+        self.proposal_id = proposal_id
+        self.state = state
+        self.ticket_count = ticket_count
+        self.rank = rank
+
+    # add a way to indicate a code has been used?
+    def use_code(self, code):
+        pass
+
+    def get_users_other_lottery_tickets_for_type(self):
+        return [
+            t
+            for t in self.user.event_tickets
+            if t.state == "entered-lottery"
+            and t != self
+            and t.proposal.type == self.proposal.type
+        ]
+
+    def is_in_lottery_round(self, round_rank):
+        return self.state == "entered-lottery" and self.rank <= round_rank
+
+    def change_state(self, new_state, *, force=False):
+        # The force option only exists for ease of resetting this in dev
+        if force:
+            self.state = new_state
+            return
+
+        if new_state == self.state:
+            return self
+
+        if new_state == "entered-lottery" and (self.rank is None or self.rank < 0):
+            raise EventTicketException("lottery tickets must have a rank")
+
+        if new_state != "entered-lottery":
+            self.rank = None
+
+        signup_state = get_signup_state()
+        valid_transitions = EVENT_TICKET_STATE_TRANSITIONS[signup_state]
+
+        try:
+            if new_state not in valid_transitions[self.state]:
+                raise EventTicketException(
+                    f"invalid state transition {self.state} -> {new_state} whilst in {signup_state}"
+                )
+        except KeyError:
+            raise EventTicketException(
+                f"State, {self.state} not found in in {signup_state}, new state {new_state}"
+            )
+
+        self.state = new_state
+        return self
+
+    def issue_codes(self):
+        # Now generate the ticket_codes
+        # These are in no way cryptographically secure etc but 1 in 308m should
+        # be low enough odds for guessing.
+        codes = []
+        for i in range(self.ticket_count):
+            codes.append("".join(choices(ascii_lowercase, k=6)))
```

Comment:
If this is going to be used, could you use something like `safechars = "2346789BCDFGHJKMPQRTVWXY` to make it less likely to make a pronounceable word?